File: Packages/TimeControlCore/Sources/TimeControlCore/Kind.swift
import Foundation

/// Category of a course or event. Fixed in code (a seeded table would duplicate under CloudKit sync).
public enum Kind: String, CaseIterable, Codable, Sendable, Identifiable {
    case course, tutorial, exam, interview, appointment, personal, other

    public var id: String { rawValue }

    public var displayName: String {
        switch self {
        case .course: "Course"
        case .tutorial: "Tutorial"
        case .exam: "Exam"
        case .interview: "Interview"
        case .appointment: "Appointment"
        case .personal: "Personal"
        case .other: "Other"
        }
    }

    public var pluralName: String {
        switch self {
        case .course: "Courses"
        case .tutorial: "Tutorials"
        case .exam: "Exams"
        case .interview: "Interviews"
        case .appointment: "Appointments"
        case .personal: "Personal"
        case .other: "Other"
        }
    }

    /// SF Symbol name.
    public var symbolName: String {
        switch self {
        case .course: "book.closed"
        case .tutorial: "person.2"
        case .exam: "pencil.and.list.clipboard"
        case .interview: "person.crop.rectangle"
        case .appointment: "calendar.badge.clock"
        case .personal: "heart"
        case .other: "tag"
        }
    }

    /// Default colour, hex RGB.
    public var colorHex: String {
        switch self {
        case .course: "#4F7CFF"
        case .tutorial: "#06B6D4"
        case .exam: "#E5484D"
        case .interview: "#F59E0B"
        case .appointment: "#10B981"
        case .personal: "#A855F7"
        case .other: "#6B7280"
        }
    }

    /// Default reminder lead time in minutes; nil = no reminder by default.
    public var defaultReminderMinutes: Int? {
        switch self {
        case .course: 10
        case .tutorial: 10
        case .exam: 60
        case .interview: 60
        case .appointment: 30
        case .personal: nil
        case .other: nil
        }
    }
}

File: web/src/core/kind.ts
/** Category of a course or event. Fixed in code, mirroring `Kind.swift`. */

export const KINDS = [
  'course',
  'tutorial',
  'exam',
  'interview',
  'appointment',
  'personal',
  'other',
] as const

export type Kind = (typeof KINDS)[number]

export function isKind(value: string): value is Kind {
  return (KINDS as readonly string[]).includes(value)
}

/** Anything the server hands us that is not a known kind reads as `other`. */
export function asKind(value: string | null | undefined): Kind {
  return value && isKind(value) ? value : 'other'
}

const DISPLAY_NAMES: Record<Kind, string> = {
  course: 'Course',
  tutorial: 'Tutorial',
  exam: 'Exam',
  interview: 'Interview',
  appointment: 'Appointment',
  personal: 'Personal',
  other: 'Other',
}

const PLURAL_NAMES: Record<Kind, string> = {
  course: 'Courses',
  tutorial: 'Tutorials',
  exam: 'Exams',
  interview: 'Interviews',
  appointment: 'Appointments',
  personal: 'Personal',
  other: 'Other',
}

const COLORS: Record<Kind, string> = {
  course: '#4F7CFF',
  tutorial: '#06B6D4',
  exam: '#E5484D',
  interview: '#F59E0B',
  appointment: '#10B981',
  personal: '#A855F7',
  other: '#6B7280',
}

const DEFAULT_REMINDERS: Record<Kind, number | null> = {
  course: 10,
  tutorial: 10,
  exam: 60,
  interview: 60,
  appointment: 30,
  personal: null,
  other: null,
}

export function kindName(kind: Kind): string {
  return DISPLAY_NAMES[kind]
}

export function kindPluralName(kind: Kind): string {
  return PLURAL_NAMES[kind]
}

export function kindColor(kind: Kind): string {
  return COLORS[kind]
}

export function kindDefaultReminderMinutes(kind: Kind): number | null {
  return DEFAULT_REMINDERS[kind]
}
